backend/scripts/location_differ.py

```diff
@@ -0,0 +1,300 @@
+"""
+This is the main script for retrieving updates from the website. It does:
+- Retrieving all countries from website (and compare to existing countries)
+- Downloading data for each country in mhtml format
+- Converting mhtml to JSON (no GPS coordinates) and build one huge JSON
+- Comparing each entry in that JSON (ID=link) to the server_json and the device_json
+- Differentiably adding the content to the server_json
+"""
+import argparse
+import json
+import os
+import logging
+
+from pennyme.locations import COUNTRY_TO_CODE, parse_location_name
+from pennyme.pennycollector import (
+    AREA_SITE,
+    AREA_PREFIX,
+    get_area_list_from_area_website,
+    validate_location_list,
+    get_prelim_geojson,
+    get_location_list_from_location_website,
+    DAY,
+    MONTH,
+    YEAR,
+    UNAVAILABLE_MACHINE_STATES,
+    get_coordinates,
+)
+from pennyme.webconfig import get_website
+from tqdm import tqdm
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "-o", "--output_folder", type=str, help="Output folder path for fresh data"
+)
+parser.add_argument(
+    "-d",
+    "--device_json",
+    type=str,
+    help="Path to the json with the machine data stored on the user device",
+)
+parser.add_argument(
+    "-s",
+    "--server_json",
+    type=str,
+    help="Path to the json with the machine data stored on the server",
+)
+parser.add_argument("-a", "--api_key", type=str, help="Google Maps API key")
+
+
+def location_differ(
+    output_folder: str, device_json: str, server_json: str, api_key: str
+):
+
+    today = f"{YEAR}-{MONTH}-{DAY}"
+
+    # Load existing json data
+    with open(device_json, "r") as f:
+        device_data = json.load(f)
+
+    with open(server_json, "r") as f:
+        server_data = json.load(f)
+
+    # Convert data to have links as keys
+    device_dict = {}
+    machine_idx = max([x["properties"]["id"] for x in device_data["features"]])
+    for geojson in device_data["features"]:
+        url = geojson["properties"]["external_url"]
+        if url not in device_dict.keys():
+            device_dict[url] = [geojson]
+        else:
+            device_dict[url].append(geojson)
+
+    device_keys = list(device_dict.keys())
+
+    server_dict = {}
+    for geojson in server_data["features"]:
+        url = geojson["properties"]["external_url"]
+        if url not in server_dict.keys():
+            server_dict[url] = [geojson]
+        else:
+            # TODO: Could be improved in the future
+            raise ValueError(
+                f"Currently, links have to be unique in server_dict, not {url}"
+            )
+        if geojson["properties"]["id"] > machine_idx:
+            machine_idx += 1
```

Comment:
to be save, we could always use the maximum, i.e. change this line to `machine_idx = geojson["properties"]["id"]`
(currently, it's only counting how many new machines are in the server data, but it would fail if any id in the server data file is not strictly sequential)

Reply:
Good point, fixed in [95b2f96](https://github.com/jannisborn/PennyMe/pull/34/commits/95b2f965df282608717dd4697f8d5a352d520a86)